src/main/java/org/broadinstitute/hellbender/utils/smithwaterman/SmithWatermanJavaAligner.java

```diff
@@ -64,37 +70,148 @@ public SmithWatermanAlignment align(final byte[] reference, final byte[] alterna
+                    alignmentResult = new SWPairwiseAlignmentResult(AlignmentUtils.consolidateCigar(new Cigar(lce)), alignmentOffset);
+
+                    Cigar cigar1 = alignmentResult.getCigar();
+                    final int n = reference.length+1;
+                    final int m = alternate.length+1;
+                    final int[][] sw = new int[n][m];
+                    final int[][] btrack=new int[n][m];
+                    calculateMatrix(reference, alternate, sw, btrack, overhangStrategy, parameters);
+                    SWPairwiseAlignmentResult alignmentResult2 = calculateCigar(sw, btrack, overhangStrategy);
+                    Cigar cigar2 = alignmentResult2.getCigar();
+                    if(!cigar1.equals(cigar2) && !cigar2.containsOperator(CigarOperator.S)){
+                        System.out.println("CIGARS NOT EQUAL");
+                        System.out.println(new String(reference));
+                        System.out.println("");
+                        System.out.println(new String(alternate));
+                        System.out.println("Heuristic:" + cigar1);
+                        System.out.println("SW:       " + cigar2);
+                    }
+                    
+                    totalComputeTime += System.nanoTime() - startTime;
+                    return alignmentResult;
+                }
+            }
+
         }
 
+        // run full Smith-Waterman
+        final int n = reference.length+1;
+        final int m = alternate.length+1;
+        final int[][] sw = new int[n][m];
+        final int[][] btrack=new int[n][m];
+
+        calculateMatrix(reference, alternate, sw, btrack, overhangStrategy, parameters);
+        alignmentResult = calculateCigar(sw, btrack, overhangStrategy); // length of the segment (continuous matches, insertions or deletions)
         totalComputeTime += System.nanoTime() - startTime;
         return alignmentResult;
     }
 
+    private static SWPairwiseAlignmentResult calculateOneIndelCigar(int indelLength, final byte[] reference, final byte[] alternate, int oneIndelIndex){
+        State state = null;
+        int cigarThirdElementLength = 0;
+
+        if(alternate.length < reference.length){
+            state = State.DELETION;
+            cigarThirdElementLength = alternate.length - oneIndelIndex;
+        }
+        if(alternate.length > reference.length){
+            state = State.INSERTION;
+            cigarThirdElementLength = alternate.length - indelLength - oneIndelIndex;
+        }
+
+        final List<CigarElement> lce = Arrays.asList(makeElement(State.MATCH, oneIndelIndex),
+                makeElement(state, indelLength), makeElement(State.MATCH, cigarThirdElementLength));
+        return new SWPairwiseAlignmentResult(AlignmentUtils.consolidateCigar(new Cigar(lce)), 0);
+    }
+
+    private static int calculateAllowedLengthOfIndelHapToRef(final SWParameters parameters){
```

Comment:
Make this public, add some tests (definitely for the SW parameters that exist in our codebase now plus maybe some novel ones to ensure it might return 0 if necessary).

Reply:
Furthermore... this seems like something that you can cache per SWParametersObject, rather than recompute these for every SW call. Though maybe they are cheap enough to forget about this...